File: java-sdk/src/main/java/com/cleanster/sdk/client/HttpClient.java
package com.cleanster.sdk.client;

import com.cleanster.sdk.exception.CleansterApiException;
import com.cleanster.sdk.exception.CleansterAuthException;
import com.cleanster.sdk.exception.CleansterException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import okhttp3.*;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

/**
 * Internal HTTP client that wraps OkHttp and handles authentication, serialization, and error handling.
 */
public class HttpClient {

    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");

    private final OkHttpClient httpClient;
    private final ObjectMapper objectMapper;
    private final CleansterConfig config;
    private volatile String bearerToken;

    public HttpClient(CleansterConfig config) {
        this.config = config;
        this.objectMapper = new ObjectMapper()
                .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

        this.httpClient = new OkHttpClient.Builder()
                .connectTimeout(config.getConnectTimeoutSeconds(), TimeUnit.SECONDS)
                .readTimeout(config.getReadTimeoutSeconds(), TimeUnit.SECONDS)
                .writeTimeout(config.getWriteTimeoutSeconds(), TimeUnit.SECONDS)
                .build();
    }

    /**
     * Set the bearer token for user-authenticated requests.
     * This token is used in the Authorization header for user-specific endpoints.
     */
    public void setBearerToken(String token) {
        this.bearerToken = token;
    }

    /**
     * Returns the currently configured bearer token (may be null).
     */
    public String getBearerToken() {
        return bearerToken;
    }

    public ObjectMapper getObjectMapper() {
        return objectMapper;
    }

    /**
     * Build the full URL for an API path.
     */
    public String url(String path) {
        return config.getBaseUrl() + path;
    }

    /**
     * Execute a GET request with access-key and optional bearer token.
     */
    public <T> T get(String path, TypeReference<T> type) {
        Request request = new Request.Builder()
                .url(url(path))
                .addHeader("Content-Type", "application/json")
                .addHeader("access-key", config.getAccessKey())
                .addHeader("token", bearerToken != null ? bearerToken : "")
                .get()
                .build();
        return execute(request, type);
    }

    /**
     * Execute a POST request with a JSON body.
     */
    public <T> T post(String path, Object body, TypeReference<T> type) {
        String json = serialize(body);
        Request request = new Request.Builder()
                .url(url(path))
                .addHeader("Content-Type", "application/json")
                .addHeader("access-key", config.getAccessKey())
                .addHeader("token", bearerToken != null ? bearerToken : "")
                .post(RequestBody.create(json, JSON))
                .build();
        return execute(request, type);
    }

    /**
     * Execute a PUT request with a JSON body.
     */
    public <T> T put(String path, Object body, TypeReference<T> type) {
        String json = serialize(body);
        Request request = new Request.Builder()
                .url(url(path))
                .addHeader("Content-Type", "application/json")
                .addHeader("access-key", config.getAccessKey())
                .addHeader("token", bearerToken != null ? bearerToken : "")
                .put(RequestBody.create(json, JSON))
                .build();
        return execute(request, type);
    }

    /**
     * Execute a DELETE request (no body).
     */
    public <T> T delete(String path, TypeReference<T> type) {
        Request request = new Request.Builder()
                .url(url(path))
                .addHeader("Content-Type", "application/json")
                .addHeader("access-key", config.getAccessKey())
                .addHeader("token", bearerToken != null ? bearerToken : "")
                .delete()
                .build();
        return execute(request, type);
    }

    /**
     * Execute a multipart POST request for file uploads.
     *
     * @param path       API path (e.g., "/v1/checklist/5/upload")
     * @param imageBytes Raw bytes of the image to upload
     * @param fileName   File name sent in the form-data part (e.g., "photo.jpg")
     * @param type       TypeReference for response deserialization
     */
    public <T> T postMultipart(String path, byte[] imageBytes, String fileName, TypeReference<T> type) {
        RequestBody fileBody = RequestBody.create(imageBytes, MediaType.parse("image/*"));
        MultipartBody requestBody = new MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("image", fileName, fileBody)
                .build();
        Request request = new Request.Builder()
                .url(url(path))
                .addHeader("access-key", config.getAccessKey())
                .addHeader("token", bearerToken != null ? bearerToken : "")
                .post(requestBody)
                .build();
        return execute(request, type);
    }

    /**
     * Execute a DELETE request with a JSON body.
     */
    public <T> T delete(String path, Object body, TypeReference<T> type) {
        String json = serialize(body);
        Request request = new Request.Builder()
                .url(url(path))
                .addHeader("Content-Type", "application/json")
                .addHeader("access-key", config.getAccessKey())
                .addHeader("token", bearerToken != null ? bearerToken : "")
                .delete(RequestBody.create(json, JSON))
                .build();
        return execute(request, type);
    }

    private <T> T execute(Request request, TypeReference<T> type) {
        try (Response response = httpClient.newCall(request).execute()) {
            String responseBody = response.body() != null ? response.body().string() : "";

            if (response.code() == 401) {
                throw new CleansterAuthException(
                        "Unauthorized - invalid or missing access key/token", responseBody);
            }
            if (!response.isSuccessful()) {
                throw new CleansterApiException(response.code(),
                        "API request failed with status " + response.code(), responseBody);
            }
            return objectMapper.readValue(responseBody, type);
        } catch (CleansterException e) {
            throw e;
        } catch (IOException e) {
            throw new CleansterException("Network error: " + e.getMessage(), e);
        }
    }

    private String serialize(Object body) {
        if (body == null) return "{}";
        try {
            return objectMapper.writeValueAsString(body);
        } catch (Exception e) {
            throw new CleansterException("Failed to serialize request body: " + e.getMessage(), e);
        }
    }
}

File: python-sdk/cleanster/http_client.py
"""
Internal HTTP transport layer — wraps the requests library and handles
authentication headers, JSON serialization, and error mapping.
"""

from typing import Any, Dict, Optional

import requests

from .config import CleansterConfig
from .exceptions import CleansterApiException, CleansterAuthException, CleansterException


class HttpClient:
    """
    Low-level HTTP client used internally by all API classes.
    Handles auth headers, timeouts, and maps HTTP error codes to SDK exceptions.
    """

    def __init__(self, config: CleansterConfig):
        self._config = config
        self._bearer_token: Optional[str] = None
        self._session = requests.Session()

    @property
    def bearer_token(self) -> Optional[str]:
        return self._bearer_token

    @bearer_token.setter
    def bearer_token(self, token: Optional[str]) -> None:
        self._bearer_token = token

    def _headers(self) -> Dict[str, str]:
        return {
            "Content-Type": "application/json",
            "access-key": self._config.access_key,
            "token": self._bearer_token or "",
        }

    def _url(self, path: str) -> str:
        return self._config.base_url + path

    def _handle_response(self, response: requests.Response) -> Dict[str, Any]:
        body = ""
        try:
            body = response.text
        except Exception:
            pass

        if response.status_code == 401:
            raise CleansterAuthException(
                "Unauthorized — invalid or missing access key or user token.",
                body,
            )
        if not response.ok:
            raise CleansterApiException(
                response.status_code,
                f"API request failed with status {response.status_code}",
                body,
            )
        try:
            return response.json()
        except Exception as exc:
            raise CleansterException(
                f"Failed to parse JSON response: {exc}"
            ) from exc

    def get(self, path: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        try:
            response = self._session.get(
                self._url(path),
                headers=self._headers(),
                params={k: v for k, v in (params or {}).items() if v is not None},
                timeout=self._config.timeout,
            )
        except requests.RequestException as exc:
            raise CleansterException(f"Network error: {exc}") from exc
        return self._handle_response(response)

    def post(self, path: str, body: Optional[Any] = None) -> Dict[str, Any]:
        try:
            response = self._session.post(
                self._url(path),
                headers=self._headers(),
                json=body,
                timeout=self._config.timeout,
            )
        except requests.RequestException as exc:
            raise CleansterException(f"Network error: {exc}") from exc
        return self._handle_response(response)

    def put(self, path: str, body: Optional[Any] = None) -> Dict[str, Any]:
        try:
            response = self._session.put(
                self._url(path),
                headers=self._headers(),
                json=body,
                timeout=self._config.timeout,
            )
        except requests.RequestException as exc:
            raise CleansterException(f"Network error: {exc}") from exc
        return self._handle_response(response)

    def post_multipart(self, path: str, image_bytes: bytes, file_name: str) -> Dict[str, Any]:
        """Upload an image via multipart/form-data POST."""
        try:
            response = self._session.post(
                self._url(path),
                headers={
                    "access-key": self._config.access_key,
                    "token": self._bearer_token or "",
                },
                files={"file": (file_name, image_bytes, "image/*")},
                timeout=self._config.timeout,
            )
        except requests.RequestException as exc:
            raise CleansterException(f"Network error: {exc}") from exc
        return self._handle_response(response)

    def delete(self, path: str, body: Optional[Any] = None) -> Dict[str, Any]:
        try:
            response = self._session.delete(
                self._url(path),
                headers=self._headers(),
                json=body,
                timeout=self._config.timeout,
            )
        except requests.RequestException as exc:
            raise CleansterException(f"Network error: {exc}") from exc
        return self._handle_response(response)
